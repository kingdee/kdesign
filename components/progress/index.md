---
category: Components
type: 反馈
order: 5
title: Progress
subtitle: 进度条
---

进度条是一种展示进度和状态的组件。

## 使用场景
- 需要展示一个操作完成的百分比时。
- 需要展示耗时任务进度时。

## API

| 参数 | 说明 | 类型 | 默认值 | 版本 |
| --- | --- | --- | --- | --- |
| className | Switch 器类名 | string | - |  |
| failureIcon | 失败时的图标 | React.ReactNode |  |  |
| infoPosition | 进度条信息位置 | string | right |  |
| percent | 当前进度数值 | number | 0 |  |
| strokeWidth | 进度条的线条粗细 | number | 8/4(type=circle) |  |
| strokeColor | 进度条的颜色（支持渐变） | string \| ProgressGradient | - |  |
| showInfo | 是否显示进度条信息，可选值：`right` `bottom` | boolean | true |  |
| status | 指定当前进度条状态，可选值：`cycle` `loading` `failure` `success` | string |  |  |
| successIcon | 成功时的图标 | React.ReactNode |  |  |
| type | 指定当前进度条类型，可选值：`line` `circle` | string | line |  |
| trailColor | 进度条未完成部分的颜色 | string | - |  |
| textMap | 进度条信息配置,值为一个数组，总共三项，依次代表加载中文字、失败文字、成功文字 | React.ReactNode[] | \['正在加载中...', '加载失败', '加载成功'\] |  |
| width | type=circle 时生效，指定当前圆形进度条的宽度 | number | 88 |  |
| format | 格式化进度条信息的方法 | Function(percent: number): React.ReactNode |  |  |
| onProcess | 进度条变化中的回调 | Function(percent: number): void |  |  |

## Design Token

| 分类 | 组件token | 全局token | 默认值 |
| --- | --- | --- | --- |
| color | --kd-c-progress-color-remaining | - | #e5e5e5 |
|  | --kd-c-progress-line-color-font | - | #212121 |
|  | --kd-c-progress-circle-color-font | - | #666 |
|  | --kd-c-progress-color-default | - | #5582f3 |
|  | --kd-c-progress-color-success | --kd-g-color-success | #1ba854 |
|  | --kd-c-progress-color-failure | --kd-g-color-error | #fb2323 |
| font | --kd-c-progress-line-icon-font-size | - | 16px |
|  | --kd-c-progress-line-text-font-size | - | 16px |
|  | --kd-c-progress-line-special-text-font-size | - | 16px |
| line-height | --kd-c-progress-circle-percent-line-height | - | 48px |
|  | --kd-c-progress-circle-unit-line-height | - | 18px |
| radius | --kd-c-progress-radius-border | - | 100px |
| sizing | --kd-c-progress-line-stroke-sizing-width | - | 8px |
|  | --kd-c-progress-circle-stroke-sizing-width | - | 4px |
| spacing | --kd-c-progress-circle-unit-spacing-margin-top | - | 18px |
|  | --kd-c-progress-line-text-spacing-margin-left | - | 8px |
|  | --kd-c-progress-line-text-spacing-margin-top | - | 12px |
|  | --kd-c-progress-circle-text-spacing-margin-top | - | 8px |
