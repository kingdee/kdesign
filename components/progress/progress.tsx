import React, { FC, useContext, useEffect } from 'react'
import classNames from 'classnames'
import Icon from '../icon'
import Line from './line'
import Circle from './circle'
import ConfigContext from '../config-provider/ConfigContext'
import { tuple } from '../_utils/type'
import { omit } from '../_utils/omit'
import { getCompProps } from '../_utils'
import { validProgress } from './utils'
import devWarning from '../_utils/devwarning'

export const ProgressTypes = tuple('line', 'circle')
export type ProgressType = typeof ProgressTypes[number]
export const ProgressStatuses = tuple('normal', 'cycle', 'loading', 'failure', 'success')
export type StringGradients = { [percentage: string]: string }
type FromToGradients = { from: string; to: string }
export type ProgressGradient = { direction?: string } & (StringGradients | FromToGradients)

export interface ProgressProps {
  type?: ProgressType
  style?: Record<string, unknown>
  className?: string
  width?: number
  strokeWidth?: number
  strokeColor?: string | ProgressGradient
  percent?: number
  showInfo?: boolean
  status?: typeof ProgressStatuses[number]
  infoPosition?: string
  trailColor?: string
  textMap?: React.ReactNode[]
  successIcon?: React.ReactNode
  failureIcon?: React.ReactNode
  format?: (percent: number) => React.ReactNode
  onProcess?: <T>(p: T) => T
  onFailedClick?: (percent: number, successPercent: number) => boolean
}

const Progress: FC<ProgressProps> = (props) => {
  const { getPrefixCls, prefixCls, compDefaultProps: userDefaultProps, locale } = useContext(ConfigContext)
  const progressProps = getCompProps('Progress', userDefaultProps, props)
  const { type, className, prefixCls: customPrefixcls, showInfo, percent, onProcess } = progressProps

  devWarning(ProgressTypes.indexOf(type) === -1, 'progress', `cannot found progress type '${type}'`)
  devWarning(type !== 'circle' && !!props.width, 'progress', 'props width only effect when type is circle')

  if (props.strokeWidth === undefined) {
    progressProps.strokeWidth = type === 'circle' ? progressProps.strokeWidth : progressProps.trailWidth
  }

  const getPercentNumber = (props: ProgressProps) => {
    const { percent = 0 } = props
    return parseInt(percent.toString(), 10)
  }

  const getProgressStatus = (props: ProgressProps) => {
    const { status } = props
    if (ProgressStatuses.indexOf(status!) < 0 && getPercentNumber(props) >= 100) {
      return 'success'
    }
    return status || 'normal'
  }

  const progressStatus = getProgressStatus(progressProps)
  const progressPrefixCls = getPrefixCls!(prefixCls, 'progress', customPrefixcls)
  const progressClasses = classNames(progressPrefixCls, className, {
    [`${progressPrefixCls}-type-${type}`]: type,
    [`${progressPrefixCls}-status-${progressStatus}`]: true,
    [`${progressPrefixCls}-show-info`]: showInfo,
    [`${progressPrefixCls}-position-bottom`]: progressProps.infoPosition !== 'right' && progressStatus === 'normal',
  })

  const renderProgressInfo = (
    prefixCls: string,
    progressStatus: typeof ProgressStatuses[number],
    props: ProgressProps,
  ) => {
    const { type = 'line', showInfo, format, percent, infoPosition, successIcon, failureIcon } = props
    const isLoading = progressStatus !== 'failure' && progressStatus !== 'success'
    if ((!showInfo && type !== 'circle') || (infoPosition !== 'right' && isLoading)) return null

    let text
    const textFormatter =
      format ||
      ((percentNumber) => (
        <>
          <span className={`${prefixCls}-text-percent`}>{percentNumber}</span>
          <span className={`${prefixCls}-text-unit`}>%</span>
        </>
      ))

    const successNodeMap = {
      line: <Icon className={`${prefixCls}-icon`} type="right-solid" />,
      circle: <Icon className={`${prefixCls}-icon`} type="right-bold" />,
      custom: successIcon,
    }
    const failureNodeMap = {
      line: <Icon className={`${prefixCls}-icon`} type="refresh-solid" />,
      circle: <Icon className={`${prefixCls}-icon`} type="close-bold" />,
      custom: failureIcon,
    }

    if (format || isLoading) {
      text = textFormatter(validProgress(percent))
    } else if (progressStatus === 'failure') {
      text = failureNodeMap.custom || failureNodeMap[type]
    } else if (progressStatus === 'success') {
      text = successNodeMap.custom || successNodeMap[type]
    }

    return <span className={`${prefixCls}-text`}>{text}</span>
  }

  const renderSpecialInfo = (
    prefixCls: string,
    progressStatus: typeof ProgressStatuses[number],
    props: ProgressProps,
  ) => {
    const { type, showInfo, infoPosition, textMap = [] } = props
    const langMsgs = locale.getCompLangMsg({ componentName: 'Progress' })
    const defaultTextMap = [langMsgs.loading, langMsgs.fail, langMsgs.success]
    const isLoading = progressStatus !== 'failure' && progressStatus !== 'success'
    if (!showInfo || (infoPosition === 'right' && type !== 'circle')) return null

    const concatMap = defaultTextMap.map((text: string, index: number) => {
      if (type === 'circle' && index === 0) return textMap[index] || langMsgs.circleLoadingDesc
      return textMap[index] || text
    })
    let text
    if (isLoading) {
      text = concatMap[0]
    } else if (progressStatus === 'failure') {
      text = concatMap[1]
    } else if (progressStatus === 'success') {
      text = concatMap[2]
    }

    return <div className={`${prefixCls}-special-text`}>{text}</div>
  }

  let progress = (
    <Line {...progressProps} prefixCls={progressPrefixCls}>
      {renderProgressInfo(progressPrefixCls, progressStatus, progressProps)}
    </Line>
  )

  if (type === 'circle') {
    progress = (
      <Circle {...progressProps} prefixCls={progressPrefixCls}>
        {renderProgressInfo(progressPrefixCls, progressStatus, progressProps)}
      </Circle>
    )
  }

  useEffect(() => {
    onProcess && onProcess(percent)
  }, [percent, onProcess])

  return (
    <div
      {...omit(progressProps, [
        'status',
        'format',
        'width',
        'strokeWidth',
        'trailWidth',
        'trailColor',
        'strokeColor',
        'percent',
        'infoPosition',
        'showInfo',
        'textMap',
        'onProcess',
      ])}
      className={progressClasses}
    >
      {progress}
      {renderSpecialInfo(progressPrefixCls, progressStatus, progressProps)}
    </div>
  )
}

Progress.displayName = 'Progress'

export default Progress
