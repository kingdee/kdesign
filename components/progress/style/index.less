@import '../../style/themes/index';
@import './mixin.less';
@progress-prefix-cls: ~'@{kd-prefix}-progress';

.@{progress-prefix-cls} {
  .progress;
  line-height: 1;
  font-size: 0;
  vertical-align: top;

  &-show-info {
    font-size: @progress-line-text-font-size;
  }

  &-type-line {
    width: 100%;
    color: @progress-line-font-color;

    .@{progress-prefix-cls}-special-text {
      margin-top: @progress-line-text-margin-top;
      line-height: 24px;
    }
  }

  &-type-circle {
    color: @progress-circle-font-color;

    .@{progress-prefix-cls}-special-text {
      margin-top: @progress-circle-text-margin-top;
      padding: 0;
      line-height: 24px;
    }
  }

  &-outer {
    display: inline-block;
    vertical-align: middle;
    box-sizing: border-box;
    width: 100%;
    margin-right: 0;
    padding-right: 0;
    font-size: 0;

    .@{progress-prefix-cls}-show-info:not(.@{progress-prefix-cls}-position-bottom) & {
      margin-right: calc(-1 * @progress-line-text-font-size * 2 - 8px);
      padding-right: calc(@progress-line-text-font-size * 2 + 8px);
    }
  }

  &-inner {
    position: relative;
    display: inline-block;
    width: 100%;
    overflow: hidden;
    vertical-align: middle;
    font-size: 0;
    border-radius: @progress-border-radius;
    background-color: @progress-remaining-color;
  }

  &-success-bg,
  &-bg {
    position: relative;
    display: inline-block;
    box-sizing: border-box;
    border-radius: @progress-border-radius;
    background-color: @progress-default-color;
    height: @progress-line-stroke-width;
    // -------------------------UPDATE-------------------
    transition: all @transition-duration @progress-transition-fn;
    // -------------------------UPDATE-------------------
  }

  &-success-bg {
    position: absolute;
    top: 0;
    left: 0;
    background-color: @progress-success-color;
  }

  &-text {
    display: inline-block;
    box-sizing: border-box;
    width: calc(@progress-line-text-font-size*2);
    margin-left: @progress-line-text-margin-left;
    // font-size: @progress-line-text-font-size;
    // line-height: 1;
    white-space: nowrap;
    text-align: left;
    vertical-align: middle;
    word-break: normal;
  }

  &-special-text {
    font-size: @progress-line-special-text-font-size;
    text-align: center;
    overflow: hidden;
    padding-right: calc(@progress-line-text-font-size*2 + 8px);

    .@{progress-prefix-cls}-position-bottom & {
      padding-right: 0;
    }
  }

  &-status-success {
    .@{progress-prefix-cls}-bg {
      background-color: @progress-success-color;
    }

    .@{progress-prefix-cls}-text,
    .@{progress-prefix-cls}-special-text {
      color: @progress-success-color;
    }

    .@{progress-prefix-cls}-text {
      font-size: 0;
    }
    .@{progress-prefix-cls}-icon {
      font-size: @progress-line-icon-font-size;
      color: @progress-success-color;
    }

    .@{progress-prefix-cls}-circle-box:not(.@{progress-prefix-cls}-circle-gradient) {
      .@{progress-prefix-cls}-circle-path {
        stroke: @progress-success-color;
      }
    }
  }

  &-status-failure {
    .@{progress-prefix-cls}-bg {
      background-color: @progress-failure-color;
    }

    .@{progress-prefix-cls}-text,
    .@{progress-prefix-cls}-special-text {
      color: @progress-failure-color;
    }

    .@{progress-prefix-cls}-text {
      font-size: 0;
    }
    .@{progress-prefix-cls}-icon {
      font-size: @progress-line-icon-font-size;
      color: @progress-failure-color;
    }

    .@{progress-prefix-cls}-circle-box:not(.@{progress-prefix-cls}-circle-gradient) {
      .@{progress-prefix-cls}-circle-path {
        stroke: @progress-failure-color;
      }
    }
  }



  &-circle-box {
    position: relative;

    .@{progress-prefix-cls}-circle-trail {
      stroke: @progress-remaining-color;
    }

    .@{progress-prefix-cls}-circle-path {
      transition: stroke-dasharray @transition-duration @progress-transition-fn, stroke-dashoffset @transition-duration @progress-transition-fn;
    }

    .@{progress-prefix-cls}-text {
      position: absolute;
      width: auto;
      text-align: center;
      font-size: @progress-circle-text-font-size;
      margin: 0;
      top: 50%;
      left: 50%;
      transform: translate(-50%, -50%);

      &-percent,
      &-unit {
        display: inline-block;
        vertical-align: top;
      }

      &-percent {
        line-height: @progress-circle-percent-line-height;
      }

      &-unit {
        line-height: @progress-circle-unit-line-height;
        font-size: @progress-circle-unit-font-size;
        margin-top: @progress-circle-unit-margin-top;
      }
    }

    .@{progress-prefix-cls}-icon {
      font-size: @progress-circle-icon-font-size;
    }
  }

  &-circle-box:not(.@{progress-prefix-cls}-circle-gradient) {
    .@{progress-prefix-cls}-circle-path {
      stroke: @progress-default-color;
      transition: stroke-dasharray @transition-duration @progress-transition-fn, stroke-dashoffset @transition-duration @progress-transition-fn;
    }
  }
}
