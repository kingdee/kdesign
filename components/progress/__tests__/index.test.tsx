import React from 'react'
import { render, mount } from 'enzyme'
import mountTest from '../../../tests/shared/mountTest'
import Progress, { ProgressType } from '../index'
import { ProgressTypes, ProgressStatuses } from '../progress'

describe('Progress', () => {
  // 1.mount test
  mountTest(Progress)

  ProgressTypes.forEach((type) => {
    mountTest(() => <Progress type={type} />)
  })

  // 2.render test
  it('renders correctly', () => {
    ProgressTypes.forEach((type) => {
      expect(render(<Progress type={type} />)).toMatchSnapshot()
    })

    ProgressStatuses.forEach((status) => {
      expect(render(<Progress status={status} />)).toMatchSnapshot()
    })
  })
  // 3.render no child without errors
  it('render no child without errors', () => {
    expect(mount(<Progress />)).toMatchSnapshot()
  })

  // 4.render null or undefined without errors
  it('render null or undefined without errors', () => {
    expect(
      mount(
        <Progress>
          {null}
          {undefined}
        </Progress>,
      ),
    ).toMatchSnapshot()
  })

  // 5.displayName
  it('should have displayName static property', () => {
    const wrapper = mount(<Progress></Progress>)
    expect((wrapper.type() as any).displayName).toBe('Progress')
  })

  // 6. warn test
  it('warns if type is wrong', () => {
    const mockWarn = jest.fn()
    jest.spyOn(console, 'warn').mockImplementation(mockWarn)
    const type = 'Wrong Type' as any as ProgressType
    render(<Progress type={type} />)
    expect(mockWarn).toHaveBeenCalledTimes(1)
    expect(mockWarn.mock.calls[0][0]).toMatch("Warning: [kdesign]-progress: cannot found progress type 'Wrong Type'")
  })

  it('warns if width use wrong', () => {
    const mockWarn = jest.fn()
    jest.spyOn(console, 'warn').mockImplementation(mockWarn)
    render(<Progress width={10} />)
    expect(mockWarn).toHaveBeenCalledTimes(1)
    expect(mockWarn.mock.calls[0][0]).toMatch(
      'Warning: [kdesign]-progress: props width only effect when type is circle',
    )
  })

  it('should class use right', () => {
    const DefaultProgress = mount(<Progress />)
    expect(DefaultProgress.find('.kd-progress')).toHaveClassName('.kd-progress-status-normal')
    expect(DefaultProgress.find('.kd-progress')).toHaveClassName('.kd-progress-show-info')

    ProgressTypes.forEach((type) => {
      const TestProgress = mount(<Progress type={type} />)
      expect(TestProgress.find('.kd-progress')).toHaveClassName(`.kd-progress-type-${type}`)
    })

    ProgressStatuses.forEach((status) => {
      const TestProgress = mount(<Progress status={status} />)
      expect(TestProgress.find('.kd-progress')).toHaveClassName(`.kd-progress-status-${status}`)
    })
  })
  // api
  it('should props width render right', () => {
    const TestProgress = mount(<Progress type="circle" width={100} />)
    expect(TestProgress.find('.kd-progress-circle-box')).toHaveStyle('width', '100px')
  })
  // percent
  it('should props percent render right', () => {
    const TestProgress = mount(<Progress percent={30} />)
    expect(TestProgress.find('.kd-progress-bg')).toHaveStyle('width', '30%')
  })
  // strokeWidth
  it('should props strokeWidth render right', () => {
    const TestProgress = mount(<Progress percent={30} strokeWidth={10} />)
    expect(TestProgress.find('.kd-progress-bg')).toHaveStyle('height', '10px')
  })
  // strokeColor
  it('should props strokeColor render right', () => {
    const TestProgress = mount(<Progress percent={30} strokeColor="#333" />)
    expect(TestProgress.find('.kd-progress-bg')).toHaveStyle('background', '#333')
  })
  // trailColor
  it('should props trailColor render right', () => {
    const TestProgress = mount(<Progress percent={30} trailColor="#333" />)
    expect(TestProgress.find('.kd-progress-inner')).toHaveStyle('backgroundColor', '#333')
  })
  // showInfo
  it('should props showInfo render right', () => {
    const LineProgress = mount(<Progress percent={30} showInfo={false} />)
    const CircleProgress = mount(<Progress type="circle" percent={30} showInfo={false} />)
    expect(LineProgress).not.toContainMatchingElement('.kd-progress-text')
    expect(CircleProgress).not.toContainMatchingElement('.kd-progress-special-text')
  })

  it("should progress's default text render right with infoPosition set", () => {
    const loadingProgress = mount(<Progress percent={30} infoPosition="bottom" />)
    const successProgress = mount(<Progress percent={100} infoPosition="bottom" />)
    const failureProgress = mount(<Progress percent={30} infoPosition="bottom" status="failure" />)
    expect(loadingProgress).toHaveText('正在加载中...')
    expect(successProgress).toHaveText('加载成功')
    expect(failureProgress).toHaveText('加载失败')
  })
  it('should props textMap render right', () => {
    const loadingDom = (
      <>
        <span style={{ float: 'left', textAlign: 'left' }}>loading...</span>
        <span style={{ float: 'right', textAlign: 'right' }}>30%</span>
      </>
    )
    const TestProgress = mount(<Progress percent={30} infoPosition="bottom" textMap={[loadingDom]} />)
    expect(TestProgress.find('.kd-progress-special-text').exists()).toBe(true)
    expect(TestProgress).toHaveText('loading...30%')
  })

  it('should gradient progress which type is circle render right', () => {
    const TestProgress = mount(
      <Progress
        type="circle"
        percent={30}
        strokeColor={{
          '0%': '#108ee9',
          '100%': '#87d068',
        }}
      />,
    )
    expect(TestProgress).toContainMatchingElement('linearGradient')
  })
  // test format
  it('should props format render right', () => {
    const TestProgress = mount(<Progress percent={30} type="circle" format={(percent) => `${percent} days`} />)
    expect(TestProgress.find('.kd-progress-text')).toHaveText('30 days')
  })
  // test onProcess callback
  it('should onProcess called when percent change', () => {
    const onProcessMock = jest.fn()
    const TestProgress = mount(<Progress percent={30} onProcess={onProcessMock} />)
    TestProgress.setProps({ percent: 40 })
    expect(onProcessMock).toHaveBeenCalled()
  })
})
