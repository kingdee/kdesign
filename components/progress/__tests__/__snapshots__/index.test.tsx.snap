// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Progress render no child without errors 1`] = `
<Progress>
  <div
    className="kd-progress kd-progress-type-line kd-progress-status-normal kd-progress-show-info"
    type="line"
  >
    <Line
      infoPosition="right"
      percent={0}
      prefixCls="kd-progress"
      showInfo={true}
      strokeWidth={8}
      trailColor="#E5E5E5"
      trailWidth={8}
      type="line"
    >
      <div
        className="kd-progress-outer"
      >
        <div
          className="kd-progress-inner"
          style={
            Object {
              "backgroundColor": "#E5E5E5",
            }
          }
        >
          <div
            className="kd-progress-bg"
            style={
              Object {
                "background": undefined,
                "height": "8px",
                "width": "0%",
              }
            }
          />
        </div>
      </div>
      <span
        className="kd-progress-text"
      >
        <span
          className="kd-progress-text-percent"
        >
          0
        </span>
        <span
          className="kd-progress-text-unit"
        >
          %
        </span>
      </span>
    </Line>
  </div>
</Progress>
`;

exports[`Progress render null or undefined without errors 1`] = `
<Progress>
  <div
    className="kd-progress kd-progress-type-line kd-progress-status-normal kd-progress-show-info"
    type="line"
  >
    <Line
      infoPosition="right"
      percent={0}
      prefixCls="kd-progress"
      showInfo={true}
      strokeWidth={8}
      trailColor="#E5E5E5"
      trailWidth={8}
      type="line"
    >
      <div
        className="kd-progress-outer"
      >
        <div
          className="kd-progress-inner"
          style={
            Object {
              "backgroundColor": "#E5E5E5",
            }
          }
        >
          <div
            className="kd-progress-bg"
            style={
              Object {
                "background": undefined,
                "height": "8px",
                "width": "0%",
              }
            }
          />
        </div>
      </div>
      <span
        className="kd-progress-text"
      >
        <span
          className="kd-progress-text-percent"
        >
          0
        </span>
        <span
          className="kd-progress-text-unit"
        >
          %
        </span>
      </span>
    </Line>
  </div>
</Progress>
`;

exports[`Progress renders correctly 1`] = `
<div
  class="kd-progress kd-progress-type-line kd-progress-status-normal kd-progress-show-info"
  type="line"
>
  <div
    class="kd-progress-outer"
  >
    <div
      class="kd-progress-inner"
      style="background-color:#E5E5E5"
    >
      <div
        class="kd-progress-bg"
        style="width:0%;height:8px"
      />
    </div>
  </div>
  <span
    class="kd-progress-text"
  >
    <span
      class="kd-progress-text-percent"
    >
      0
    </span>
    <span
      class="kd-progress-text-unit"
    >
      %
    </span>
  </span>
</div>
`;

exports[`Progress renders correctly 2`] = `
<div
  class="kd-progress kd-progress-type-circle kd-progress-status-normal kd-progress-show-info"
  type="circle"
>
  <div
    class="kd-progress-circle-box"
    style="width:88px;height:88px"
  >
    <svg
      class="kd-progress-circle"
      stroke-width="4"
      viewBox="0 0 100 100"
    >
      <path
        class="kd-progress-circle-trail"
        d="M 50,50 m 0,-48 

                      a 48,48 0 1 1 0,96 

                      a 48,48 0 1 1 0,-96 
"
        fill-opacity="0"
        stroke-linecap="round"
        stroke-width="4"
        style="stroke:#E5E5E5"
      />
      <path
        class="kd-progress-circle-path"
        d="M 50,50 m 0,-48 

                      a 48,48 0 1 1 0,96 

                      a 48,48 0 1 1 0,-96 
"
        fill-opacity="0"
        stroke-linecap="round"
        stroke-width="4"
        style="stroke-dasharray:0px 301.59289474462014px;stroke-dashoffset:0px;opacity:0"
      />
    </svg>
    <span
      class="kd-progress-text"
    >
      <span
        class="kd-progress-text-percent"
      >
        0
      </span>
      <span
        class="kd-progress-text-unit"
      >
        %
      </span>
    </span>
  </div>
  <div
    class="kd-progress-special-text"
  >
    正在加载中
  </div>
</div>
`;

exports[`Progress renders correctly 3`] = `
<div
  class="kd-progress kd-progress-type-line kd-progress-status-normal kd-progress-show-info"
  type="line"
>
  <div
    class="kd-progress-outer"
  >
    <div
      class="kd-progress-inner"
      style="background-color:#E5E5E5"
    >
      <div
        class="kd-progress-bg"
        style="width:0%;height:8px"
      />
    </div>
  </div>
  <span
    class="kd-progress-text"
  >
    <span
      class="kd-progress-text-percent"
    >
      0
    </span>
    <span
      class="kd-progress-text-unit"
    >
      %
    </span>
  </span>
</div>
`;

exports[`Progress renders correctly 4`] = `
<div
  class="kd-progress kd-progress-type-line kd-progress-status-cycle kd-progress-show-info"
  type="line"
>
  <div
    class="kd-progress-outer"
  >
    <div
      class="kd-progress-inner"
      style="background-color:#E5E5E5"
    >
      <div
        class="kd-progress-bg"
        style="width:0%;height:8px"
      />
    </div>
  </div>
  <span
    class="kd-progress-text"
  >
    <span
      class="kd-progress-text-percent"
    >
      0
    </span>
    <span
      class="kd-progress-text-unit"
    >
      %
    </span>
  </span>
</div>
`;

exports[`Progress renders correctly 5`] = `
<div
  class="kd-progress kd-progress-type-line kd-progress-status-loading kd-progress-show-info"
  type="line"
>
  <div
    class="kd-progress-outer"
  >
    <div
      class="kd-progress-inner"
      style="background-color:#E5E5E5"
    >
      <div
        class="kd-progress-bg"
        style="width:0%;height:8px"
      />
    </div>
  </div>
  <span
    class="kd-progress-text"
  >
    <span
      class="kd-progress-text-percent"
    >
      0
    </span>
    <span
      class="kd-progress-text-unit"
    >
      %
    </span>
  </span>
</div>
`;

exports[`Progress renders correctly 6`] = `
<div
  class="kd-progress kd-progress-type-line kd-progress-status-failure kd-progress-show-info"
  type="line"
>
  <div
    class="kd-progress-outer"
  >
    <div
      class="kd-progress-inner"
      style="background-color:#E5E5E5"
    >
      <div
        class="kd-progress-bg"
        style="width:0%;height:8px"
      />
    </div>
  </div>
  <span
    class="kd-progress-text"
  >
    <i
      class="kdicon kd-progress-icon kdicon-refresh-solid"
    />
  </span>
</div>
`;

exports[`Progress renders correctly 7`] = `
<div
  class="kd-progress kd-progress-type-line kd-progress-status-success kd-progress-show-info"
  type="line"
>
  <div
    class="kd-progress-outer"
  >
    <div
      class="kd-progress-inner"
      style="background-color:#E5E5E5"
    >
      <div
        class="kd-progress-bg"
        style="width:0%;height:8px"
      />
    </div>
  </div>
  <span
    class="kd-progress-text"
  >
    <i
      class="kdicon kd-progress-icon kdicon-right-solid"
    />
  </span>
</div>
`;
